canonical/maas-ui
# docs: add storybook

## Done

- add storybook with example stories for `AppSideNavigation`, `Meter`, `TooltipButon`, `MachineActionFormWrapper`
- include `@storybook/addon-a11y` for easy accessibility testing
- refactor AppSideNavigation
- extend `HACKING.md` docs
- export context providers from `AppRoot` as `RootProviders` (so that they can be used in storybook)

## Rationale
This PR adds storybook to maas-ui to encourage [component-driven](https://www.componentdriven.org/) development and good component design.

It allows to quickly preview various states of components acting as an interactive documentation and speeding up the development.

It's not using extensive customisation options or specifying separate variants for each type of component to keep things simple and minimize the maintenance burden.

For example, MachineActionFormWrapper is not added as separate Stories for each action (deploy, abort, etc.). This would have the benefit of them being displayed in the navigation on the left, but would require specifying each story manually. We can achieve the same thing by simply switching the "action" prop using storybook controls. These are added automatically and don't require any extra code.

## What's not included
- Mocking of the API.
- Customisation of redux store/dispatching actions (will be added in the future). For that reason some of the machine actions may render unexpected results like an infinitely loading spinner.

## QA

### QA steps

- run yarn storybook
- verify you can view components and control their props via storybook controls


## Screenshots

### After
![Google Chrome screenshot 000820@2x](https://github.com/canonical/maas-ui/assets/7452681/3e37593e-92d2-4448-8111-cf1d0c2bcefe)
![Google Chrome screenshot 000834@2x](https://github.com/canonical/maas-ui/assets/7452681/8c6d9fd5-315e-4edc-97ef-4cb093912138)

![Google Chrome screenshot 000828@2x](https://github.com/canonical/maas-ui/assets/7452681/1120d1d9-02e4-4d2b-99e5-b7fbf23c8ff2)
